[dynamo][guards] Skip no tensor aliasing guards on parameters

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139038
* #138942
* __->__ #138954
* #139040

This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with one parameter input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

anijain2305 committed:
[dynamo][guards] Skip no tensor aliasing guards on parameters

This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with the same input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip no tensor aliasing guards on parameters"


This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with the same input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip no tensor aliasing guards on parameters"


This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with the same input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ezyang commented:
I am sympathetic to this but it doesn't seem like it should be taking 40us to test this in the first place

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -1850,14 +1837,23 @@ def TENSOR_MATCH(self, guard: Guard, value=None):
             else:
-                self.tensor_check_examples.append(value)
-                self.tensor_check_names.append(tensor_name)
-                self.tensor_check_guards.append(guard)
```
There's a refactor going on here that makes it hard to see the substantive change that happened here

anijain2305 replied:
I can pull the refactor in a separate PR.

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -2300,17 +2296,17 @@ def add_code_part(code_part, guard, log_only=False):
                     add_code_part(code, gcl.guard, True)
                     seen.add(code)
 
-        tensor_check_names = builder.tensor_check_names
```
Does the code in torch/csrc/dynamo/guards.cpp referencing tensor_check_names have to be adjusted?

Also, since this happened before, can you explicitly make sure our verbose reporting (e.g., in logs and tlparse) did not regress with these changes? (The switch to cpp guard manager caused us to lose all tlparse guard output)

anijain2305 commented:
> I am sympathetic to this but it doesn't seem like it should be taking 40us to test this in the first place

The quantized model in question has a really large number of parameters, and therefore we end up calling this guard for every parameter. There is not much going on in the guard. Its just the sheer quantity of them.

anijain2305 committed:
Update on "[dynamo][guards] Skip no tensor aliasing guards on parameters"


This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with one parameter input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip no tensor aliasing guards on parameters"


This is another unsound guard eval optimization. Its rare in practice to
compile a function with two different parameters as inputs, and then
later call the function with one parameter input as two different inputs
(aliasing). This further reduces guard overhead from 280 us to 240 us
for the model in https://github.com/pytorch/pytorch/issues/138386

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]
